README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# equationer <img src="man/figures/logo.png" align="right" height=140/>

[![Build
Status](https://travis-ci.com/UBESP-DCTV/equationer.svg?branch=master)](https://travis-ci.com/UBESP-DCTV/equationer)
[![Coverage
status](https://codecov.io/gh/UBESP-DCTV/equationer/branch/master/graph/badge.svg)](https://codecov.io/github/UBESP-DCTV/equationer?branch=master)

[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![CRAN
status](https://www.r-pkg.org/badges/version/equationer)](https://cran.r-project.org/package=equationer)

The goal of the package `equationer` is to implement an environment to
consistently generate the estimations from the equations reported in the
paper “Review of Equations of Energy Requirements” by the *Unit of
Biostatistics, Epidemiology and Public Health* of the University of
Padova.

## Installation

You can install the development version of `equationer` from
[GitHub](https://github.com/) with the following procedure:

``` r
# install.packages("devtools")
devtools::install_github("UBESP-DCTV/equationer")
```

## Hot to use it

``` r
library(equationer)
```

First of all define some equations. Please note that equations must be
named and that equations’ names must be different each other. Moreover,
you have to supply strata as a named list defining in each equation
which level of the strata it consider. Note that strata levels can be of
different class inside an equation (e.g. character, numeric, …) but,
obviously, you cannot provide multiple strata with the same name or a
strata with more than one level in each equation.

``` r
eq1 <- eq(age = 0.3, bmi = -0.5, name = "eq-1", outcome = "kcal/day",
    strata = list(sex = "male", nyha = 1)
)
eq2 <- eq(age = 0.5, bmi = -0.3, name = "eq-2", outcome = "kcal/day",
    strata = list(sex = "female", nyha = 1)
)
eq3 <- eq(age = -0.3, bmi = 0.5, name = "eq-3", outcome = "kcal/day",
    strata = list(sex = "male", nyha = 2)
)
eq4 <- eq(age = -0.5, bmi = 0.3, name = "eq-4", outcome = "kcal/day",
    strata = list(sex = "female", nyha = 2)
)


eq9 <- eq(age = -0.1, weight = 0.2, name = "eq-9", outcome = "kcal/month",
    strata = list(sex = "female")
)
eq10 <- eq(age = -0.2, weight = 0.1, name = "eq-10", outcome = "kcal/month",
    strata = list(sex = "male")
)
eq11 <- eq(age = 0.1, weight = -0.2, name = "eq-11", outcome = "kcal/day",
    strata = list(sex = "female")
)
eq12 <- eq(age = 0.2, weight = -0.2, name = "eq-12", outcome = "kcal/day",
    strata = list(sex = "male")
)
```

When you visualize an equation the main characteristics will be
highlighted

``` r
eq1
#> 
#> ● Equation 'eq-1': kcal/day = 0.3 age + -0.5 bmi
#> Strata: sex = 'male'
#> Strata: nyha = 1
eq2
#> 
#> ● Equation 'eq-2': kcal/day = 0.5 age + -0.3 bmi
#> Strata: sex = 'female'
#> Strata: nyha = 1
eq3
#> 
#> ● Equation 'eq-3': kcal/day = -0.3 age + 0.5 bmi
#> Strata: sex = 'male'
#> Strata: nyha = 2

eq9
#> 
#> ● Equation 'eq-9': kcal/month = -0.1 age + 0.2 weight
#> Strata: sex = 'female'
```

We can evaluate an equation providing (in a natural non quoted language)
the values for each . Please note that we have to provide “at least” the
covariate included in the equation (or we get an error that tell us
which ones we have missed). If we provide more covariates than the ones
included in the equation it does not matter for the resulting estimation
and its value will be returned in the output data frame too.

``` r
evaluate_at(eq1, age = 35)
#> ✖ Covariate(s) missing: 'bmi'
#> Error: Not all equation's covariates are included in the set supplied.
evaluate_at(eq1, age = 35, bmi = 18)
#> # A tibble: 1 x 7
#>     age   bmi sex    nyha outcome  estimation eq_name
#>   <dbl> <dbl> <chr> <dbl> <chr>         <dbl> <chr>  
#> 1    35    18 male      1 kcal/day        1.5 eq-1
evaluate_at(eq9, age = 35, bmi = 18, weight = 84)
#> # A tibble: 1 x 6
#>     age weight sex    outcome    estimation eq_name
#>   <dbl>  <dbl> <chr>  <chr>           <dbl> <chr>  
#> 1    35     84 female kcal/month       13.3 eq-9
```

Equations must be aggregated in group, basically because the strata
which cannot suit well in maths: i.e. each different combination of
strata (i.e. categorical level) there is an equation. For consistency,
even single equation with no strata must be grouped in a group
(singleton). Even for the groups of equations we have to provide
(distinct) names for each group. Moreover, you can optionally associate
a reference (i.e. a single character string) to groups.

When we print a group, a summary of the characteristics were shown
including each equation itself

``` r
eqs1 <- eqs(eq1, eq2, eq3, eq4, name = "eqs-a", reference = "ref-a")
eqs2 <- eqs(eq9, eq10, eq11, eq12, name = "eqs-b", reference = "ref-b")

eqs1
#> 
#> Equations group 'eqs-a':
#> 
#> Strata sex w/ levels: 'female', 'male'
#> Strata nyha w/ levels: '1', '2'
#> 
#> 
#> ● Equation 'eq-1': kcal/day = 0.3 age + -0.5 bmi
#> Strata: sex = 'male'
#> Strata: nyha = 1
#> 
#> 
#> ● Equation 'eq-2': kcal/day = 0.5 age + -0.3 bmi
#> Strata: sex = 'female'
#> Strata: nyha = 1
#> 
#> 
#> ● Equation 'eq-3': kcal/day = -0.3 age + 0.5 bmi
#> Strata: sex = 'male'
#> Strata: nyha = 2
#> 
#> 
#> ● Equation 'eq-4': kcal/day = -0.5 age + 0.3 bmi
#> Strata: sex = 'female'
#> Strata: nyha = 2
#> 
#> 
#> Main reference: 'ref-a'
```

We can evaluate a group of equation all together as it is a single one.
We will get a data frame of results with one row each possible equation.
Possible equation were decided with respect to the outcome requested
(default all the possible one), i.e. if some equation in a group cannot
provide an outcome, clearly those equation will not evaluate it. At the
same time we can considered for the evaluation only equations which
assume particular level of some strata. In this case if we ask for a
stratum to assume a level it does not have nothing will be computed
(anyway the resulting object still a data frame of zero rows, for
consistency of the output class). On the other hand if we ask for a
strata that does not exist at all in the equation in the group it will
be ignored in the computation and the level selected will be reported in
the output data frame.

``` r
# not enough values
evaluate_at(eqs1, age = 38)
#> ✖ Covariate(s) missing: 'bmi'
#> Error: Not all equation's covariates are included in the set supplied.

# enough values, no filter
evaluate_at(eqs1, age = 38, bmi = 18)
#> # A tibble: 4 x 9
#>     age   bmi sex     nyha outcome  estimation eq_name eq_group reference
#>   <dbl> <dbl> <chr>  <dbl> <chr>         <dbl> <chr>   <chr>    <chr>    
#> 1    38    18 male       1 kcal/day        2.4 eq-1    eqs-a    ref-a    
#> 2    38    18 female     1 kcal/day       13.6 eq-2    eqs-a    ref-a    
#> 3    38    18 male       2 kcal/day       -2.4 eq-3    eqs-a    ref-a    
#> 4    38    18 female     2 kcal/day      -13.6 eq-4    eqs-a    ref-a

# more values, no filter
evaluate_at(eqs1, age = 38, bmi = 18, weight = 81, height = 184)
#> # A tibble: 4 x 9
#>     age   bmi sex     nyha outcome  estimation eq_name eq_group reference
#>   <dbl> <dbl> <chr>  <dbl> <chr>         <dbl> <chr>   <chr>    <chr>    
#> 1    38    18 male       1 kcal/day        2.4 eq-1    eqs-a    ref-a    
#> 2    38    18 female     1 kcal/day       13.6 eq-2    eqs-a    ref-a    
#> 3    38    18 male       2 kcal/day       -2.4 eq-3    eqs-a    ref-a    
#> 4    38    18 female     2 kcal/day      -13.6 eq-4    eqs-a    ref-a

# filter by a strata level
evaluate_at(eqs1, age = 38, bmi = 18, sex = "female")
#> # A tibble: 2 x 9
#>     age   bmi sex     nyha outcome  estimation eq_name eq_group reference
#>   <dbl> <dbl> <chr>  <dbl> <chr>         <dbl> <chr>   <chr>    <chr>    
#> 1    38    18 female     1 kcal/day       13.6 eq-2    eqs-a    ref-a    
#> 2    38    18 female     2 kcal/day      -13.6 eq-4    eqs-a    ref-a

# filter by a strata nonexistent level
evaluate_at(eqs1, age = 38, bmi = 18, sex = "unknown")
#> Warning: Some strata's levels requested are not included in the equations in
#> eqs-a. They won't be evaluated.
#> ✖ Values 'unknown' is not present in the sex's strata possible levels.
#> ● Possible values for sex are: 'female', 'male'.
#> # A tibble: 0 x 9
#> # … with 9 variables: age <dbl>, bmi <dbl>, sex <fct>, nyha <fct>,
#> #   outcome <chr>, estimation <dbl>, eq_name <chr>, eq_group <chr>,
#> #   reference <chr>

# Ask only for a specific outcome (please note the DOT)
evaluate_at(eqs2, age = 38, weight = 81, .outcome = "kcal/day")
#> # A tibble: 2 x 8
#>     age weight sex    outcome  estimation eq_name eq_group reference
#>   <dbl>  <dbl> <chr>  <chr>         <dbl> <chr>   <chr>    <chr>    
#> 1    38     81 female kcal/day     -12.4  eq-11   eqs-b    ref-b    
#> 2    38     81 male   kcal/day      -8.60 eq-12   eqs-b    ref-b

# Ask only for a specific nonexistent outcome
evaluate_at(eqs2, age = 38, weight = 81, .outcome = "kcal")
#> Warning: Only equations with possible outcome (if any) will be considered for
#> 'eqs-b'.
#> ✖ Outcome 'kcal' is not present in the equations eqs-b
#> ● Possible outcomes are: 'kcal/month', 'kcal/day'.
#> # A tibble: 0 x 8
#> # … with 8 variables: age <dbl>, weight <dbl>, sex <fct>, outcome <chr>,
#> #   estimation <dbl>, eq_name <chr>, eq_group <chr>, reference <chr>
```

Finally, a bag of (grouped) equations can be created to have a complete
bag\!

``` r
bag1 <- eqs_bag(eqs1, eqs2, name = "ubesp19",
    reference = "DG et.al 2019"
)

bag1
#> 
#> Equations bag 'ubesp19' (last update `2020-07-08`)
#> 
#> Variable included: 'age', 'bmi', 'weight'
#> Strata: sex w/ levels: 'female', 'male'
#> Strata: nyha w/ levels: 1, 2
#> Outcome considered: 'kcal/day', 'kcal/month'
#> Number of equation groups: 2
#> Overall numeber of equations: 8
#> 
#> Main reference: 'DG et.al 2019'
```

For a complete bag (which is supposed to be quite huge) we can ask an
estimation both by providing directly the covariates value (i.e. one
patient by one patient), and by passing the information (of one or more
patients) using a data frame\! In the latter case, a column which keep
track of the original row in the sourced dataframe will be added

``` r
evaluate_at(bag1,
    age = 35, bmi = 18, weight = 81, sex = "female"
)
#> Warning: Only equations with possible outcome (if any) will be considered for
#> 'eqs-a'.
#> ✖ Outcome 'kcal/month' is not present in the equations eqs-a
#> ● Possible outcomes are: 'kcal/day'.
#> # A tibble: 4 x 10
#>     age   bmi sex     nyha outcome  estimation eq_name eq_group reference weight
#>   <dbl> <dbl> <chr>  <dbl> <chr>         <dbl> <chr>   <chr>    <chr>      <dbl>
#> 1    35    18 female     1 kcal/day       12.1 eq-2    eqs-a    ref-a         NA
#> 2    35    18 female     2 kcal/day      -12.1 eq-4    eqs-a    ref-a         NA
#> 3    35    NA female    NA kcal/mo…       12.7 eq-9    eqs-b    ref-b         81
#> 4    35    NA female    NA kcal/day      -12.7 eq-11   eqs-b    ref-b         81

one_patient <- dplyr::tribble(
    ~age, ~bmi, ~weight,     ~sex,
      35,   18,      81,  "female"
)

evaluate_at(bag1, one_patient)
#> Warning: Only equations with possible outcome (if any) will be considered for
#> 'eqs-a'.
#> ✖ Outcome 'kcal/month' is not present in the equations eqs-a
#> ● Possible outcomes are: 'kcal/day'.
#> # A tibble: 4 x 10
#>     age   bmi sex     nyha outcome  estimation eq_name eq_group reference weight
#>   <dbl> <dbl> <chr>  <dbl> <chr>         <dbl> <chr>   <chr>    <chr>      <dbl>
#> 1    35    18 female     1 kcal/day       12.1 eq-2    eqs-a    ref-a         NA
#> 2    35    18 female     2 kcal/day      -12.1 eq-4    eqs-a    ref-a         NA
#> 3    35    NA female    NA kcal/mo…       12.7 eq-9    eqs-b    ref-b         81
#> 4    35    NA female    NA kcal/day      -12.7 eq-11   eqs-b    ref-b         81



more_patients <- dplyr::tribble(
    ~age, ~bmi, ~weight,     ~sex,
      35,   18,      81,  "female",
      27,   20,      93,    "male"
)

evaluate_at(bag1, more_patients)
#> Warning: Only equations with possible outcome (if any) will be considered for
#> 'eqs-a'.
#> ✖ Outcome 'kcal/month' is not present in the equations eqs-a
#> ● Possible outcomes are: 'kcal/day'.
#> Warning: Only equations with possible outcome (if any) will be considered for
#> 'eqs-a'.
#> ✖ Outcome 'kcal/month' is not present in the equations eqs-a
#> ● Possible outcomes are: 'kcal/day'.
#> # A tibble: 8 x 11
#>     age   bmi sex    nyha outcome estimation eq_name eq_group reference weight
#>   <dbl> <dbl> <chr> <dbl> <chr>        <dbl> <chr>   <chr>    <chr>      <dbl>
#> 1    35    18 fema…     1 kcal/d…       12.1 eq-2    eqs-a    ref-a         NA
#> 2    35    18 fema…     2 kcal/d…      -12.1 eq-4    eqs-a    ref-a         NA
#> 3    35    NA fema…    NA kcal/m…       12.7 eq-9    eqs-b    ref-b         81
#> 4    35    NA fema…    NA kcal/d…      -12.7 eq-11   eqs-b    ref-b         81
#> 5    27    20 male      1 kcal/d…       -1.9 eq-1    eqs-a    ref-a         NA
#> 6    27    20 male      2 kcal/d…        1.9 eq-3    eqs-a    ref-a         NA
#> 7    27    NA male     NA kcal/m…        3.9 eq-10   eqs-b    ref-b         93
#> 8    27    NA male     NA kcal/d…      -13.2 eq-12   eqs-b    ref-b         93
#> # … with 1 more variable: .source_row <int>
```

## Code of Conduct

Please note that the `equationer` project is released with a
[Contributor Code of Conduct](.github/CODE_OF_CONDUCT.md). By
contributing to this project, you agree to abide by its terms.


NEWS.md
# equationer 0.6.1

* fix locale-specific error message in tests
* changed Anthropometric to Demographic (#22)

# equationer 0.4.1

* Fix `usethis::` declarations

# equationer 0.4.0

* Update equations from HO (still remain many commented out to be fixed)

# equationer 0.3.3

* Hot fix ibw calculation
* Adjusted weight calculated automatically in the shiny app

# equationer (development version)

* Hot-fix a sign error in a equation (#17)

# equationer 0.3.2

* Updated text in the shiny app

# equationer 0.3.1

* fix a bug that prevent the join of the partial tables if full strata
  are provided and some partial results were evaluated without any
  real results (0-row df)
* hotfix wrong coefficient in Muller04-a

# equationer 0.3.0

* updated the Shiny interface
* bugfix converting logical to character in the `reer` data to prevent a
  joining error
* removed computable covariates from shiny (and internally compute and
  use them showing output value if useful)
* bugfix for intercept considered internally as strata 
* implemented and checked all the equations (#5)

# equationer (development version)

* Hotfix for menopausal in shiny that still constant

# equationer 0.2.2

* restyle shiny UI 
* update shiny text for usage explanation
* bugfix errors and messages in shiny (#13)

# equationer 0.2.1

* Shiny messages bug fixes

# equationer 0.2.0

* added warning if not sufficient input are supplied (#11)
* remove the needs to provide the intercept
* fix bug found by DB that prevents `evaluate_at()` to work if
  no strata are provided to a bag (#10)
* fix bug that prevents `evaluate_at()` to work if data supplied 
  strata not in the eqs_bag.

# equationer 0.1.3

* fix bug which prevent `evaluate_at()` to works with bag which include
  equations group which differ in variables that match the same regexpr,
  e.g, "weight" and "weight_adjusted" both match the regexpr "weight".

# equationer 0.1.2

* fix a bug which prevents `eqs()` to create a set of equations if
  they have no strata at all.

# equationer 0.1.1

* Updated README 
* print methods.

# equationer 0.1.0

* Support for data frame in put in `evaluate_at.eqs_bag()`
* Added `eqs_bag` class, constructor, extractors, and main methods,
  to collect equations from the same "reference" and evaluate them
  together;
* Updated tests moving all the data in `test/testthat/helper-eq.R`.

# equationer (development version)

* Added `eqs` class including constructor, extractors and main methods,
  to collect equations from the same "reference" and evaluate them
  together;
* Changed strata attributes to list of factors;
* Added support for tidy evaluation (`rlang` pkg main utilities).

# equationer 0.0.0.9001

* Added basic methods for class `eq`; 
* Added extractors for class `eq`;
* Added `eq()` constructor for object of class `eq`;
* Added ui and utils functions;
* Insert Logo;
* update Travis.

# equationer 0.0.0.9000

* Added basic development support:
  - Contributing, issue template, support, and CoC information;
  - git + GitHub;
  - appVeyor + Travis + codecov;
  - gpl3 license;
  - testthat + roxygen2 + spellcheck;
  - `` magrittr::`%>%` `` + `tibble::tibble`;
  - `README.Rmd` + `README.md`.

* Added a `NEWS.md` file to track changes to the package.


R/evaluate_at.R
#' Evaluate equation
#'
#' Evaluate (i.e., solve for the outcome) the supplied equation at the
#' supplied values of the covariates
#'
#' @param x equation(s set/bag) object
#' @param ... sequence of named coefficients' values
#'
#' @return a one row [tibble][tibble::tibble-package] with one column
#'     each covariate supplied, one column for possible strata
#'     considered, one column with the outcome, one column with
#'     the value of the outcome that solve the equation when
#'     its the covariates assume the values supplied, and one last
#'     column with the name of the equation used
#' @export
evaluate_at <- function(x, ...) {
    UseMethod("evaluate_at", x)
}

#' @describeIn evaluate_at Evaluate (i.e., solve for the outcome) the
#'     supplied equation at the supplied values of the covariates.
#' @export
#'
#' @examples
#' library(equationer)
#' eq_test <- eq(age = 0.1, bmi = -0.3,
#'     name    = "first eq_test",
#'     outcome = "out"
#' )
#' eq2_test <- eq(age = 0.1, bmi = -0.3,
#'     name    = "first eq_test",
#'     outcome = "out",
#'     strata = list(sex = "female")
#' )
#'
#' evaluate_at(eq_test, age = 38, bmi = 18)
#' evaluate_at(eq_test, age = 38, sex = 1, bmi = 18)
#' evaluate_at(eq2_test, age = 38, bmi = 18)
#'
#' \dontrun{
#'     # ERRORS
#'     evaluate_at(eq_test, age = 38, bmi = 18, sex = "female")
#'     evaluate_at(eq_test, age = 38)
#' }
evaluate_at.eq <- function(x, ...) {

    vs <- list(...)
    eq_cov <- get_covariates(x)

    vs <- vs[names(vs) %in% eq_cov]

    if (!rlang::is_named(vs)) {
        ui_stop("Not all variable names are valid or non empty names")
    }

    if (any(duplicated(names(vs)))) {
        ui_stop("Some variable names are duplicated.")
    }



    are_cov_num <- purrr::map_lgl(vs, is.numeric)
    if (!all(are_cov_num)) {
        not_num <- vs[!are_cov_num]
        ui_stop(c(
            "All covariates values must be numeric.",
            ui_fail("Not numeric values: {ui_value(not_num)}")
        ))
    }

    vs[["intercept"]] <- 1L

    cov_val_name <- names(vs)

    if (!is_applicable_to_covset(x, cov_val_name)) {
        missing_covs <- eq_cov %>%
            setdiff(cov_val_name)

        ui_stop(c(
            "Not all equation's covariates are included in the set supplied.",
            ui_fail("Covariate(s) missing: {ui_value(missing_covs)}")
        ))
    }



    res <- unlist(vs[eq_cov]) %*% x %>%
        drop() %>%
        purrr::set_names(get_outcome(x))

    res <- dplyr::as_tibble(c(
        vs,
        get_strata(x),
        list(
            outcome = get_outcome(x),
            estimation = res,
            eq_name = get_name(x)
        )
    ))

    res[["intercept"]] <- NULL # not select in case it not exists
    res
}







#' @describeIn evaluate_at Evaluate (i.e., solve for the outcome)
#'     all the \code{\link{eq}}uations in the \code{\link{eqs}} object
#'     at the supplied values of the covariates.
#'
#' @param .outcome (chr, default NULL) an optional declaration of which
#'     outcome(s) the computation has to consider
#' @export
#'
#' @examples
#' library(equationer)
#' eq_test <- eq(age = 0.5, bmi = -0.3,
#'     name    = "eq_test",
#'     outcome = "kcal/day",
#'     strata = list(sex = "female")
#' )
#' eq2_test <- eq(age = 0.3, bmi = -0.5,
#'     name    = "eq2_test",
#'     outcome = "kcal/day",
#'     strata = list(sex = "male")
#' )
#'
#' eqs_test <- eqs(eq_test, eq2_test, name = "eqs-test")
#'
#' evaluate_at(eqs_test, age = 38, bmi = 18, sex = "female")
#' evaluate_at(eqs_test, age = 38, bmi = 18, sex = "male")
#'
#' \dontrun{
#'     # ERRORS
#'     evaluate_at(eq_test, age = 38, bmi = 18, mellitus = "yes"))
#'     evaluate_at(eqs_test, age = 38, bmi = 18, sex = "unknown")
#'     evaluate_at(eq_test, age = 38)
#' }
evaluate_at.eqs <- function(x, ..., .outcome = NULL) {
    vs <- list(...)

    if (!rlang::is_named(vs)) {
        ui_stop("Not all variable/strata names are valid or non empty names.")
    }

    vs[["intercept"]] <- 1L
    vs_names <- names(vs)
    if (any(duplicated(vs_names))) {
        ui_stop("Some variable/strata names are duplicated.")
    }

    x_strata <- get_strata(x)
    x_strata_names <- names(x_strata)


    vs_strata_values <- vs[vs_names %in% x_strata_names]
    are_values_in_levels <- purrr::imap_lgl(vs_strata_values, ~{
        .x %in% levels(x_strata[[.y]])
    })
    if (!all(are_values_in_levels)) {
        ui_warn("Some strata's levels requested are not included in the equations in {ui_field(get_name(x))}. They won't be evaluated.")
        purrr::iwalk(vs_strata_values[!are_values_in_levels], ~{
            ui_fail("Values {ui_value(.x)} is not present in the {ui_field(.y)}'s strata possible levels.")
            ui_todo("Possible values for {ui_field(.y)} are: {ui_value(levels(x_strata[[.y]]))}.")
        })
    }

    .outcome <- .outcome %||% get_outcome(x)

    are_possible_outcome <- .outcome %in% get_outcome(x)

    if (!all(are_possible_outcome)) {
        ui_warn("Only equations with possible outcome (if any) will be considered for {ui_value(get_name(x))}.")
        purrr::walk(.outcome[!are_possible_outcome], ~{
            ui_fail("Outcome {ui_value(.x)} is not present in the equations {ui_field(get_name(x))}")
            ui_todo("Possible outcomes are: {ui_value(unique(get_outcome(x)))}.")
        })
    }

    strata_df <- c(
            id = list(seq_along(x)),
            x_strata,
            outcome = list(get_outcome(x))
        ) %>%
        dplyr::as_tibble() %>%
        dplyr::filter(!!rlang::sym("outcome") %in% .outcome)

    purrr::iwalk(vs_strata_values, ~{
        strata_name <- sym(.y)

        strata_df <<- strata_df %>%
            dplyr::filter(!!strata_name == .x)
    })

    equation_to_use <- strata_df[["id"]]

    vs_covs_values <- vs[!vs_names %in% x_strata_names]

    res <- purrr::map_df(equation_to_use,
        ~do.call(evaluate_at, c(list(x = x[[.x]]), vs_covs_values))
    )

    if (!nrow(res)) {
        res <- dplyr::as_tibble(
            c(
                purrr::map(vs_covs_values, ~.x[0]),
                purrr::map(x_strata, ~.x[0]),
                list(
                    outcome = character(),
                    estimation = numeric(),
                    eq_name = character()
                )
            ),
            .rows = 0
        )
    }

    res <- dplyr::as_tibble(c(
        res,
        list(
            eq_group = get_name(x),
            reference = get_reference(x)
        )
    ))
    res[["intercept"]] <- NULL
    res

}









#' @describeIn evaluate_at Evaluate (i.e., solve for the outcome)
#'     all the \code{\link{eq}}uations in all the possible
#'     \code{\link{eqs}} at the supplied values of the covariates.
#'
#' @export
#'
#' @examples
#'
#' eq1 <- eq(age = 0.3, bmi = -0.5,
#'     name    = "cl_test_1",
#'     outcome = "kcal/day",
#'     strata = list(sex = "male", nyha = 1)
#' )
#'
#' eq2 <- eq(age = 0.5, bmi = -0.3,                 # change strata wrt eq1
#'     name = "cl_test_2",
#'     outcome = "kcal/day",
#'     strata = list(sex = "female", nyha = 1)
#' )
#'
#' eq3 <- eq(age = -0.3, bmi = 0.5,
#'     name    = "cl_test_3",
#'     outcome = "kcal/day",
#'     strata = list(sex = "male", nyha = 2)
#' )
#' eq4 <- eq(age = -0.5, bmi = 0.3,
#'     name = "cl_test_4",
#'     outcome = "kcal/day",
#'     strata = list(sex = "female", nyha = 2)
#' )
#'
#' eq9 <- eq(age = -0.1, weight = 0.2,                # change var, strata
#'     name = "cl_test_9",
#'     outcome = "kcal/month",
#'     strata = list(sex = "female")
#' )
#'
#' eq10 <- eq(age = -0.2, weight = 0.1,                # change var, strata
#'     name = "cl_test_10",
#'     outcome = "kcal/month",
#'     strata = list(sex = "male")
#' )
#'
#' eq11 <- eq(age = 0.1, weight = -0.2,                # change var, strata
#'     name = "cl_test_11",
#'     outcome = "kcal/day",
#'     strata = list(sex = "female")
#' )
#'
#' eq12 <- eq(age = 0.2, weight = -0.2,                # change var, strata
#'     name = "cl_test_12",
#'     outcome = "kcal/day",
#'     strata = list(sex = "male")
#' )
#'
#'
#' eqs1 <- eqs(eq1, eq2, eq3, eq4,
#'     name = "eqs1",
#'     reference = "ref-a"
#' )
#'
#' eqs2 <- eqs(eq9, eq10, eq11, eq12,
#'     name = "eqs2",
#'     reference = "ref-b"
#' )
#'
#'
#' eqs_bag_test <- eqs_bag(eqs1, eqs2,
#'     name = "overall-bag",
#'     reference = "equationer-test-bag"
#' )
#'
#'
#' evaluate_at(eqs_bag_test, age = 35)
#' evaluate_at(eqs_bag_test, age = 35, bmi = 18)
#' evaluate_at(eqs_bag_test, age = 35, bmi = 18, weight = 81)
#' evaluate_at(eqs_bag_test, age = 35, bmi = 18, sex = "female")
#' evaluate_at(eqs_bag_test, age = 35, bmi = 18, weight = 81, sex = "female")
#' evaluate_at(eqs_bag_test, age = 35, bmi = 18, weight = 81, nyha = 1)
#' evaluate_at(eqs_bag_test, age = 35, bmi = 18, weight = 81, .outcome = "kcal/day")
#' evaluate_at(eqs_bag_test, age = 35, bmi = 18, weight = 81, sex = "female", .outcome = "kcal/day")
#'
#' one_patient <- dplyr::tribble(
#'     ~id, ~age, ~bmi, ~weight,     ~sex,
#'       1,   35,   18,      81,  "female"
#' )
#'
#' more_patients <- dplyr::tribble(
#'     ~id, ~age, ~bmi, ~weight,     ~sex,
#'       1,   35,   18,      81,  "female",
#'       2,   27,   20,      93,    "male"
#' )
#'
#' evaluate_at(eqs_bag_test, one_patient)
#' evaluate_at(eqs_bag_test, more_patients)
evaluate_at.eqs_bag <- function(x, ..., .outcome = NULL) {
    vs <- list(...)

    if (length(vs) == 1) {

        single_vs <- vs[[1]]

        if (inherits(single_vs, "data.frame") && nrow(single_vs) < 2) {
            var_lst <- as.list(single_vs)
            res <- do.call(evaluate_at,
                c(var_lst, list(x = x, .outcome = .outcome))
            )
            return(res)
        }

        if (inherits(single_vs, "data.frame")) {
            res_dfs <- purrr::map(seq_len(nrow(single_vs)), ~{
                evaluate_at(
                    x = x,
                    single_vs[.x, , drop = FALSE],
                    .outcome = .outcome
                ) %>%
                    dplyr::mutate(.source_row = .x)
            })
            res <- suppressMessages(
                purrr::reduce(res_dfs, dplyr::full_join)
            )
            return(res)
        }
    }



    if (!rlang::is_named(vs)) {
        ui_stop("Not all variable/strata names are valid or non empty names.")
    }

    vs[["intercept"]] <- 1L
    vs_names <- names(vs)
    if (any(duplicated(vs_names))) {
        ui_stop("Some variable/strata names are duplicated.")
    }

    x_strata <- get_strata(x)
    x_strata_names <- names(x_strata)

    vs_strata_values <- vs[vs_names %in% x_strata_names]

    are_values_in_levels <- purrr::imap_lgl(vs_strata_values, ~{
        .x %in% x_strata[[.y]]
    })

    if (!all(are_values_in_levels)) {
        missing_strata <- vs_strata_values[!are_values_in_levels]
        ui_warn("Some strata's levels requested are not included in the equations in {ui_field(get_name(x))}.\n Only equation without {ui_field(names(missing_strata))} strata will be evaluated.")
        purrr::iwalk(missing_strata, ~{
            ui_fail("Values {ui_value(.x)} is not present in the levels of {ui_field(.y)} for equations in {ui_value(get_name(x))}.")
            ui_todo("Possible values for {ui_field(.y)} are: {ui_value(x_strata[[.y]])}.")
        })
    }

    .outcome <- .outcome %||% get_outcome(x)
    are_possible_outcome <- .outcome %in% get_outcome(x)

    if (!all(are_possible_outcome)) {
        ui_warn("(Some) outcome requested is not evaluated using equations in {ui_value(get_name(x))}. They will be excluded")
        purrr::walk(.outcome[!are_possible_outcome], ~{
            ui_fail("Outcome {ui_value(.x)} is not present in the equations of the group {ui_field(get_name(x))}")
            ui_todo("Possible outcomes for {ui_field(get_name(x))} are: {ui_value(unique(get_outcome(x)))}.")
        })
    }

    vs_covs_values <- vs[!vs_names %in% x_strata_names]
    vs_covs_names  <- names(vs_covs_values)

    vs_strata_values <- vs_strata_values[are_values_in_levels]

    .outcome <- .outcome[are_possible_outcome]

    eqs_to_consider <- which_are_applicable_to_covset(x, vs_covs_names)

    if (!length(eqs_to_consider)) {
        res <- dplyr::as_tibble(
            c(
                purrr::map(vs_covs_values, ~.x[0]),
                purrr::map(x_strata, ~.x[0]),
                list(
                    outcome = character(),
                    estimation = numeric(),
                    eq_name = character(),
                    eq_group = character(),
                    reference = character()
                )
            ),
            .rows = 0
        )
        res[["intercept"]] <- NULL
        return(res)
    }

    reduced_bag <- do.call(eqs_bag,
        c(
            x[eqs_to_consider],
            list(
                name = get_name(x),
                reference = get_reference(x),
                last_update = get_last_update(x)
            )
        )
    )

    res_lst <- purrr::map(reduced_bag,
        ~do.call(evaluate_at, c(
            vs_covs_values,
            vs_strata_values,
            list(
                x = .x,
                .outcome = .outcome
            )
        ))
    )

    if (length(res_lst) == 1) {
        res_lst[[1]][["intercept"]] <- NULL
        return(res_lst[[1]])
    }

    res_lst <- res_lst[purrr::map_lgl(res_lst, ~nrow(.x) > 0L)]

    suppressMessages(
        res <- purrr::reduce(res_lst, dplyr::full_join)
    )
    res[["intercept"]] <- NULL
    res
}

